File: apps/backend/src/routes/dashboards.ts
import { Router } from 'express';
import { z } from 'zod';
import { pool } from '../db/client.js';

const router = Router();

const LLM_SERVICE_URL = process.env.LLM_SERVICE_URL ?? 'http://localhost:8000';
const DEFAULT_LLM_TIMEOUT_MS = 180_000;

function readPositiveIntEnv(name: string, fallback: number): number {
  const raw = process.env[name];
  if (!raw) return fallback;
  const parsed = Number(raw);
  return Number.isFinite(parsed) && parsed > 0 ? parsed : fallback;
}

const LLM_TIMEOUT_MS = readPositiveIntEnv('LLM_TIMEOUT_MS', DEFAULT_LLM_TIMEOUT_MS);

const GenerateSchema = z.object({
  prompt:        z.string().min(3),
  resourceIds:   z.array(z.string().uuid()).default([]),
  docsUrls:      z.array(z.string()).default([]),
  variantCount:  z.number().int().min(1).max(8).default(4),
});

// ─── Schemas ──────────────────────────────────────────────────────────────────

const ConfigSchema = z
  .object({
    components: z.array(z.any()),
    queries: z.array(z.any()),
  })
  .passthrough();

const CreateSchema = z.object({
  name:   z.string().min(1, 'name is required'),
  slug:   z.string().optional(),
  config: ConfigSchema,
  status: z.enum(['draft', 'live']).default('draft'),
});

const UpdateSchema = z.object({
  name:   z.string().min(1).optional(),
  slug:   z.string().optional(),
  config: ConfigSchema.optional(),
  status: z.enum(['draft', 'live']).optional(),
});

// ─── Helpers ──────────────────────────────────────────────────────────────────

function toSlug(name: string): string {
  return name
    .toLowerCase()
    .trim()
    .replace(/[^a-z0-9\s-]/g, '')
    .replace(/\s+/g, '-')
    .replace(/-+/g, '-')
    .replace(/^-|-$/g, '');
}

const PG_UNIQUE_VIOLATION    = '23505'; // duplicate key
const PG_INVALID_UUID        = '22P02'; // invalid UUID text representation

function pgCode(err: unknown): string | undefined {
  return (err as any)?.code;
}

// ─── POST /api/dashboards ─────────────────────────────────────────────────────

router.post('/', async (req, res) => {
  const parsed = CreateSchema.safeParse(req.body);
  if (!parsed.success) {
    return res.status(400).json({
      error: 'Validation failed',
      details: parsed.error.flatten(),
    });
  }

  const { name, config, status } = parsed.data;
  let slug = parsed.data.slug ?? toSlug(name);
  let attempts = 0;

  while (attempts < 10) {
    try {
      const { rows } = await pool.query<DashboardRow>(
        `INSERT INTO dashboards (name, slug, config, status)
         VALUES ($1, $2, $3, $4)
         RETURNING id, name, slug, config, status, created_at, updated_at`,
        [name, slug, config, status],
      );
      return res.status(201).json(rows[0]);
    } catch (err) {
      if (pgCode(err) === PG_UNIQUE_VIOLATION) {
        attempts++;
        slug = `${parsed.data.slug ?? toSlug(name)}-${attempts}`;
        continue;
      }
      console.error('[dashboards] create:', err);
      return res.status(500).json({ error: 'Internal server error' });
    }
  }

  return res.status(409).json({ error: 'Could not generate a unique slug after 10 attempts' });
});

// ─── GET /api/dashboards ──────────────────────────────────────────────────────

router.get('/', async (_req, res) => {
  try {
    // Pull assigned customers from BOTH sources:
    //   1. The new many-to-many `dashboard_assignments` table (post-migration)
    //   2. The legacy 1-to-1 `customers.dashboard_id` column (pre-migration data)
    // Without the legacy union, customers created before the many-to-many work
    // appear as "No customer assigned" on the dashboard cards.
    const { rows } = await pool.query<any>(
      `SELECT
         d.id, d.name, d.slug, d.status, d.created_at, d.updated_at,
         COALESCE(
           (
             SELECT jsonb_agg(jsonb_build_object('id', sub.id, 'name', sub.name, 'slug', sub.slug))
             FROM (
               SELECT c.id, c.name, c.slug
               FROM customers c
               JOIN dashboard_assignments da
                 ON da.customer_id = c.id AND da.dashboard_id = d.id
               UNION
               SELECT c.id, c.name, c.slug
               FROM customers c
               WHERE c.dashboard_id = d.id
             ) sub
           ),
           '[]'::jsonb
         ) AS assigned_customers
       FROM dashboards d
       ORDER BY d.updated_at DESC, d.created_at DESC`,
    );
    return res.json(rows);
  } catch (err) {
    console.error('[dashboards] list:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
});

// ─── POST /api/dashboards/generate ────────────────────────────────────────────
// Proxy to the Python LLM microservice. Loads the requested resources +
// their imported endpoints, packs the LLM payload, and forwards. Does NOT
// save anything to the dashboards table — caller (template-picker UI)
// decides which variant to persist.

router.post('/generate', async (req, res) => {
  const parsed = GenerateSchema.safeParse(req.body);
  if (!parsed.success) {
    return res.status(400).json({
      error:   'Validation failed',
      details: parsed.error.flatten(),
    });
  }
  const { prompt, resourceIds, docsUrls, variantCount } = parsed.data;

  // Hydrate the resource IDs into the shape the LLM service expects.
  let resourcesPayload: Array<{
    id:        string;
    name:      string;
    type:      string;
    base_url:  string | null;
    endpoints: Array<{ method: string; path: string; summary: string | null }>;
  }> = [];

  if (resourceIds.length > 0) {
    try {
      const { rows } = await pool.query<{
        id:        string;
        name:      string;
        type:      string;
        base_url:  string | null;
        endpoints: Array<{ method: string; path: string; summary: string | null }> | null;
      }>(
        `SELECT
           r.id, r.name, r.type, r.base_url,
           COALESCE(
             jsonb_agg(
               jsonb_build_object('method', e.method, 'path', e.path, 'summary', e.summary)
               ORDER BY e.path, e.method
             ) FILTER (WHERE e.id IS NOT NULL),
             '[]'::jsonb
           ) AS endpoints
         FROM resources r
         LEFT JOIN resource_endpoints e ON e.resource_id = r.id
         WHERE r.id = ANY($1::uuid[])
         GROUP BY r.id`,
        [resourceIds],
      );
      resourcesPayload = rows.map((r) => ({
        id:        r.id,
        name:      r.name,
        type:      r.type,
        base_url:  r.base_url,
        endpoints: r.endpoints ?? [],
      }));
    } catch (err) {
      console.error('[dashboards] generate resource lookup:', err);
      return res.status(500).json({ error: 'Could not load resources for the LLM' });
    }
  }

  // Forward to the Python service with a hard timeout so a stuck Gemini
  // call doesn't hang the dashboard backend.
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), LLM_TIMEOUT_MS);

  try {
    const response = await fetch(`${LLM_SERVICE_URL}/generate`, {
      method:  'POST',
      headers: { 'Content-Type': 'application/json' },
      body:    JSON.stringify({
        prompt,
        resources:    resourcesPayload,
        docsUrls,
        variantCount,
      }),
      signal: controller.signal,
    });

    const text = await response.text();
    let json: unknown = null;
    try { json = JSON.parse(text); } catch { /* keep raw text for error path */ }

    if (!response.ok) {
      const detail = (json && typeof json === 'object' && 'detail' in json)
        ? String((json as Record<string, unknown>).detail)
        : text || `LLM service returned ${response.status}`;
      return res.status(502).json({ error: detail });
    }

    return res.json(json);
  } catch (err) {
    if ((err as Error).name === 'AbortError') {
      const seconds = Math.round(LLM_TIMEOUT_MS / 1000);
      return res.status(504).json({ error: `LLM service took longer than ${seconds} seconds` });
    }
    console.error('[dashboards] generate proxy error:', err);
    return res.status(502).json({ error: 'Could not reach the LLM service' });
  } finally {
    clearTimeout(timer);
  }
});

// ─── GET /api/dashboards/:id ──────────────────────────────────────────────────

router.get('/:id', async (req, res) => {
  try {
    const { rows } = await pool.query<DashboardRow>(
      `SELECT id, name, slug, config, status, created_at, updated_at
       FROM dashboards
       WHERE id = $1`,
      [req.params.id],
    );
    if (rows.length === 0) {
      return res.status(404).json({ error: 'Dashboard not found' });
    }
    return res.json(rows[0]);
  } catch (err) {
    if (pgCode(err) === PG_INVALID_UUID) {
      return res.status(400).json({ error: 'Invalid dashboard ID' });
    }
    console.error('[dashboards] get:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
});

// ─── PUT /api/dashboards/:id ──────────────────────────────────────────────────

router.put('/:id', async (req, res) => {
  const parsed = UpdateSchema.safeParse(req.body);
  if (!parsed.success) {
    return res.status(400).json({
      error: 'Validation failed',
      details: parsed.error.flatten(),
    });
  }

  const data = parsed.data;

  if (Object.keys(data).length === 0) {
    return res.status(400).json({ error: 'No fields provided to update' });
  }

  // Build dynamic SET clause from whichever fields were supplied.
  // Slug is derived from the incoming name when name changes but slug is omitted.
  const setClauses: string[] = [];
  const values: unknown[]    = [];
  let   i = 1;

  if (data.name !== undefined) {
    setClauses.push(`name = $${i++}`);
    values.push(data.name);
  }

  const newSlug = data.slug ?? (data.name !== undefined ? toSlug(data.name) : undefined);
  if (newSlug !== undefined) {
    setClauses.push(`slug = $${i++}`);
    values.push(newSlug);
  }

  if (data.config !== undefined) {
    setClauses.push(`config = $${i++}`);
    values.push(data.config);
  }

  if (data.status !== undefined) {
    setClauses.push(`status = $${i++}`);
    values.push(data.status);
  }

  setClauses.push('updated_at = NOW()');
  values.push(req.params.id);

  try {
    const { rows } = await pool.query<DashboardRow>(
      `UPDATE dashboards
       SET ${setClauses.join(', ')}
       WHERE id = $${i}
       RETURNING id, name, slug, config, status, created_at, updated_at`,
      values,
    );
    if (rows.length === 0) {
      return res.status(404).json({ error: 'Dashboard not found' });
    }
    return res.json(rows[0]);
  } catch (err) {
    if (pgCode(err) === PG_UNIQUE_VIOLATION) {
      return res.status(409).json({ error: 'Slug is already taken' });
    }
    if (pgCode(err) === PG_INVALID_UUID) {
      return res.status(400).json({ error: 'Invalid dashboard ID' });
    }
    console.error('[dashboards] update:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
});

// ─── PATCH /api/dashboards/:id/publish ────────────────────────────────────────

router.patch('/:id/publish', async (req, res) => {
  const schema = z.object({
    status: z.enum(['draft', 'live']),
  });

  const parsed = schema.safeParse(req.body);
  if (!parsed.success) {
    return res.status(400).json({ error: 'Invalid status' });
  }

  const { status } = parsed.data;
  const publishedAt = status === 'live' ? 'NOW()' : 'NULL';

  try {
    const { rows } = await pool.query<DashboardRow>(
      `UPDATE dashboards
       SET status = $1, published_at = ${publishedAt}, updated_at = NOW()
       WHERE id = $2
       RETURNING id, name, slug, config, status, published_at, created_at, updated_at`,
      [status, req.params.id],
    );

    if (rows.length === 0) {
      return res.status(404).json({ error: 'Dashboard not found' });
    }

    return res.json(rows[0]);
  } catch (err) {
    if (pgCode(err) === PG_INVALID_UUID) {
      return res.status(400).json({ error: 'Invalid dashboard ID' });
    }
    console.error('[dashboards] publish:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
});

// ─── POST /api/dashboards/:id/assign ──────────────────────────────────────────

router.post('/:id/assign', async (req, res) => {
  const schema = z.object({
    customer_ids: z.array(z.string().uuid()),
  });

  const parsed = schema.safeParse(req.body);
  if (!parsed.success) {
    return res.status(400).json({
      error: 'Validation failed',
      details: parsed.error.flatten(),
    });
  }

  const { customer_ids } = parsed.data;
  const dashboardId = req.params.id;

  try {
    // 1. Validate dashboard exists
    const { rows: dashRows } = await pool.query('SELECT id FROM dashboards WHERE id = $1', [dashboardId]);
    if (dashRows.length === 0) {
      return res.status(404).json({ error: 'Dashboard not found' });
    }

    // 2. Sync assignments (delete existing, then insert new)
    await pool.query('BEGIN');
    try {
      await pool.query('DELETE FROM dashboard_assignments WHERE dashboard_id = $1', [dashboardId]);
      
      if (customer_ids.length > 0) {
        // Use a single query for multiple inserts if possible, or loop for simplicity
        for (const cid of customer_ids) {
          await pool.query(
            `INSERT INTO dashboard_assignments (dashboard_id, customer_id)
             VALUES ($1, $2)`,
            [dashboardId, cid]
          );
        }
      }
      await pool.query('COMMIT');
    } catch (e) {
      await pool.query('ROLLBACK');
      throw e;
    }

    // 3. Return updated list of assigned customers
    const { rows: assignedCustomers } = await pool.query(
      `SELECT c.id, c.name, c.slug
       FROM customers c
       JOIN dashboard_assignments da ON da.customer_id = c.id
       WHERE da.dashboard_id = $1
       ORDER BY c.name ASC`,
      [dashboardId]
    );

    return res.json(assignedCustomers);
  } catch (err) {
    if (pgCode(err) === PG_INVALID_UUID) {
      return res.status(400).json({ error: 'Invalid ID' });
    }
    console.error('[dashboards] assign:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
});

// ─── GET /api/dashboards/:id/customers ────────────────────────────────────────

router.get('/:id/customers', async (req, res) => {
  try {
    const { rows } = await pool.query(
      `SELECT c.id, c.name, c.slug
       FROM customers c
       JOIN dashboard_assignments da ON da.customer_id = c.id
       WHERE da.dashboard_id = $1
       ORDER BY c.name ASC`,
      [req.params.id]
    );
    return res.json(rows);
  } catch (err) {
    if (pgCode(err) === PG_INVALID_UUID) {
      return res.status(400).json({ error: 'Invalid dashboard ID' });
    }
    console.error('[dashboards] get assigned customers:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
});

// ─── DELETE /api/dashboards/:id ───────────────────────────────────────────────

router.delete('/:id', async (req, res) => {
  try {
    const result = await pool.query(
      'DELETE FROM dashboards WHERE id = $1',
      [req.params.id],
    );
    if (!result.rowCount) {
      return res.status(404).json({ error: 'Dashboard not found' });
    }
    return res.status(204).send();
  } catch (err) {
    if (pgCode(err) === PG_INVALID_UUID) {
      return res.status(400).json({ error: 'Invalid dashboard ID' });
    }
    if (pgCode(err) === '23503') {
      return res.status(409).json({ error: 'Dashboard is still referenced — unassign any customers first' });
    }
    console.error('[dashboards] delete:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
});

// ─── Row types (for pg generic typing) ───────────────────────────────────────

interface DashboardRow {
  id:           string;
  name:         string;
  slug:         string;
  config:       unknown;
  status:       string;
  published_at: Date | null;
  created_at:   Date;
  updated_at:   Date;
}

interface DashboardSummaryRow {
  id:         string;
  name:       string;
  slug:       string;
  status:     string;
  created_at: Date;
  updated_at: Date;
}

export default router;

File: apps/backend-py/app/routes/dashboards.py
import asyncio
import json
import re
from typing import Any

import asyncpg
import httpx
from fastapi import APIRouter, Depends, HTTPException, Request
from pydantic import BaseModel, field_validator

from app.config import settings
from app.db.client import get_pool
from app.middleware.auth import require_auth

router = APIRouter()

PG_UNIQUE_VIOLATION = "23505"
PG_INVALID_UUID = "22P02"
PG_FOREIGN_KEY = "23503"

UUID_RE = re.compile(r"^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$", re.IGNORECASE)


def _to_slug(name: str) -> str:
    s = name.lower().strip()
    s = re.sub(r"[^a-z0-9\s-]", "", s)
    s = re.sub(r"\s+", "-", s)
    s = re.sub(r"-+", "-", s)
    return s.strip("-")


def _pg_code(exc: Exception) -> str | None:
    return getattr(exc, "pgcode", None) or getattr(exc, "sqlstate", None)


class DashboardConfig(BaseModel):
    components: list[Any]
    queries: list[Any]
    model_config = {"extra": "allow"}


class CreateDashboard(BaseModel):
    name: str
    slug: str | None = None
    config: DashboardConfig
    status: str = "draft"

    @field_validator("status")
    @classmethod
    def validate_status(cls, v: str) -> str:
        if v not in ("draft", "live"):
            raise ValueError("status must be 'draft' or 'live'")
        return v


class UpdateDashboard(BaseModel):
    name: str | None = None
    slug: str | None = None
    config: DashboardConfig | None = None
    status: str | None = None

    @field_validator("status")
    @classmethod
    def validate_status(cls, v: str | None) -> str | None:
        if v is not None and v not in ("draft", "live"):
            raise ValueError("status must be 'draft' or 'live'")
        return v


class GenerateRequest(BaseModel):
    prompt: str
    resourceIds: list[str] = []
    docsUrls: list[str] = []
    variantCount: int = 4

    @field_validator("prompt")
    @classmethod
    def prompt_min(cls, v: str) -> str:
        if len(v) < 3:
            raise ValueError("prompt must be at least 3 characters")
        return v

    @field_validator("variantCount")
    @classmethod
    def variant_range(cls, v: int) -> int:
        if not (1 <= v <= 8):
            raise ValueError("variantCount must be between 1 and 8")
        return v


@router.post("", status_code=201, dependencies=[Depends(require_auth)])
async def create_dashboard(body: CreateDashboard, pool: asyncpg.Pool = Depends(get_pool)):
    slug = body.slug or _to_slug(body.name)
    base_slug = slug

    for attempt in range(10):
        try:
            row = await pool.fetchrow(
                """INSERT INTO dashboards (name, slug, config, status)
                   VALUES ($1, $2, $3::jsonb, $4)
                   RETURNING id, name, slug, config, status, created_at, updated_at""",
                body.name,
                slug,
                body.config.model_dump(),
                body.status,
            )
            return dict(row)
        except asyncpg.UniqueViolationError:
            attempt_n = attempt + 1
            slug = f"{base_slug}-{attempt_n}"
        except Exception as exc:
            raise HTTPException(status_code=500, detail="Internal server error") from exc

    raise HTTPException(status_code=409, detail="Could not generate a unique slug after 10 attempts")


@router.get("", dependencies=[Depends(require_auth)])
async def list_dashboards(pool: asyncpg.Pool = Depends(get_pool)):
    rows = await pool.fetch(
        """SELECT
             d.id, d.name, d.slug, d.status, d.created_at, d.updated_at,
             COALESCE(
               (
                 SELECT jsonb_agg(jsonb_build_object('id', sub.id, 'name', sub.name, 'slug', sub.slug))
                 FROM (
                   SELECT c.id, c.name, c.slug
                   FROM customers c
                   JOIN dashboard_assignments da ON da.customer_id = c.id AND da.dashboard_id = d.id
                   UNION
                   SELECT c.id, c.name, c.slug
                   FROM customers c
                   WHERE c.dashboard_id = d.id
                 ) sub
               ),
               '[]'::jsonb
             ) AS assigned_customers
           FROM dashboards d
           ORDER BY d.updated_at DESC, d.created_at DESC"""
    )
    return [dict(r) for r in rows]


@router.post("/generate", dependencies=[Depends(require_auth)])
async def generate_dashboard(body: GenerateRequest, pool: asyncpg.Pool = Depends(get_pool)):
    resources_payload: list[dict] = []

    if body.resourceIds:
        rows = await pool.fetch(
            """SELECT
                 r.id, r.name, r.type, r.base_url,
                 COALESCE(
                   jsonb_agg(
                     jsonb_build_object('method', e.method, 'path', e.path, 'summary', e.summary)
                     ORDER BY e.path, e.method
                   ) FILTER (WHERE e.id IS NOT NULL),
                   '[]'::jsonb
                 ) AS endpoints
               FROM resources r
               LEFT JOIN resource_endpoints e ON e.resource_id = r.id
               WHERE r.id = ANY($1::uuid[])
               GROUP BY r.id""",
            body.resourceIds,
        )
        resources_payload = [
            {
                "id": str(r["id"]),
                "name": r["name"],
                "type": r["type"],
                "base_url": r["base_url"],
                "endpoints": r["endpoints"] or [],
            }
            for r in rows
        ]

    timeout_s = settings.llm_timeout_ms / 1000
    llm_url = settings.llm_service_url.rstrip("/")

    try:
        async with httpx.AsyncClient(timeout=timeout_s) as client:
            resp = await client.post(
                f"{llm_url}/generate",
                json={
                    "prompt": body.prompt,
                    "resources": resources_payload,
                    "docsUrls": body.docsUrls,
                    "variantCount": body.variantCount,
                },
            )
        if not resp.is_success:
            try:
                detail = resp.json().get("detail") or resp.text
            except Exception:
                detail = resp.text
            raise HTTPException(status_code=502, detail=str(detail))
        return resp.json()
    except httpx.TimeoutException:
        secs = int(timeout_s)
        raise HTTPException(status_code=504, detail=f"LLM service took longer than {secs} seconds")
    except HTTPException:
        raise
    except Exception as exc:
        raise HTTPException(status_code=502, detail="Could not reach the LLM service") from exc


@router.get("/{dashboard_id}", dependencies=[Depends(require_auth)])
async def get_dashboard(dashboard_id: str, pool: asyncpg.Pool = Depends(get_pool)):
    try:
        row = await pool.fetchrow(
            "SELECT id, name, slug, config, status, created_at, updated_at FROM dashboards WHERE id = $1::uuid",
            dashboard_id,
        )
    except asyncpg.InvalidTextRepresentationError:
        raise HTTPException(status_code=400, detail="Invalid dashboard ID")
    except Exception as exc:
        raise HTTPException(status_code=500, detail="Internal server error") from exc

    if not row:
        raise HTTPException(status_code=404, detail="Dashboard not found")
    return dict(row)


@router.put("/{dashboard_id}", dependencies=[Depends(require_auth)])
async def update_dashboard(dashboard_id: str, body: UpdateDashboard, pool: asyncpg.Pool = Depends(get_pool)):
    data = body.model_dump(exclude_none=True)
    if not data:
        raise HTTPException(status_code=400, detail="No fields provided to update")

    base_slug: str | None = body.slug or (_to_slug(body.name) if body.name else None)
    current_slug = base_slug

    for attempt in range(10):
        set_clauses: list[str] = []
        values: list[Any] = []
        i = 1

        if body.name is not None:
            set_clauses.append(f"name = ${i}")
            values.append(body.name)
            i += 1
        if current_slug is not None:
            set_clauses.append(f"slug = ${i}")
            values.append(current_slug)
            i += 1
        if body.config is not None:
            set_clauses.append(f"config = ${i}::jsonb")
            values.append(body.config.model_dump())
            i += 1
        if body.status is not None:
            set_clauses.append(f"status = ${i}")
            values.append(body.status)
            i += 1

        set_clauses.append("updated_at = NOW()")
        values.append(dashboard_id)

        try:
            row = await pool.fetchrow(
                f"""UPDATE dashboards SET {', '.join(set_clauses)}
                    WHERE id = ${i}::uuid
                    RETURNING id, name, slug, config, status, created_at, updated_at""",
                *values,
            )
        except asyncpg.UniqueViolationError:
            if base_slug:
                current_slug = f"{base_slug}-{attempt + 1}"
                continue
            raise HTTPException(status_code=409, detail="Slug is already taken")
        except asyncpg.InvalidTextRepresentationError:
            raise HTTPException(status_code=400, detail="Invalid dashboard ID")
        except Exception as exc:
            raise HTTPException(status_code=500, detail="Internal server error") from exc

        if not row:
            raise HTTPException(status_code=404, detail="Dashboard not found")
        return dict(row)

    raise HTTPException(status_code=409, detail="Could not generate a unique slug after 10 attempts")


@router.patch("/{dashboard_id}/publish", dependencies=[Depends(require_auth)])
async def publish_dashboard(dashboard_id: str, request: Request, pool: asyncpg.Pool = Depends(get_pool)):
    try:
        body = await request.json()
    except Exception:
        raise HTTPException(status_code=400, detail="Invalid JSON")

    status = body.get("status")
    if status not in ("draft", "live"):
        raise HTTPException(status_code=400, detail="Invalid status")

    published_at_expr = "NOW()" if status == "live" else "NULL"

    try:
        row = await pool.fetchrow(
            f"""UPDATE dashboards
                SET status = $1, published_at = {published_at_expr}, updated_at = NOW()
                WHERE id = $2::uuid
                RETURNING id, name, slug, config, status, published_at, created_at, updated_at""",
            status,
            dashboard_id,
        )
    except asyncpg.InvalidTextRepresentationError:
        raise HTTPException(status_code=400, detail="Invalid dashboard ID")
    except Exception as exc:
        raise HTTPException(status_code=500, detail="Internal server error") from exc

    if not row:
        raise HTTPException(status_code=404, detail="Dashboard not found")
    return dict(row)


@router.post("/{dashboard_id}/assign", dependencies=[Depends(require_auth)])
async def assign_customers(dashboard_id: str, request: Request, pool: asyncpg.Pool = Depends(get_pool)):
    try:
        body = await request.json()
    except Exception:
        raise HTTPException(status_code=400, detail="Invalid JSON")

    customer_ids: list[str] = body.get("customer_ids", [])
    if not isinstance(customer_ids, list):
        raise HTTPException(status_code=400, detail="customer_ids must be an array")

    try:
        async with pool.acquire() as conn:
            async with conn.transaction():
                # Validate dashboard exists
                dash_exists = await conn.fetchval("SELECT 1 FROM dashboards WHERE id = $1::uuid", dashboard_id)
                if not dash_exists:
                    raise HTTPException(status_code=404, detail="Dashboard not found")

                await conn.execute("DELETE FROM dashboard_assignments WHERE dashboard_id = $1::uuid", dashboard_id)
                for cid in customer_ids:
                    await conn.execute(
                        "INSERT INTO dashboard_assignments (dashboard_id, customer_id) VALUES ($1::uuid, $2::uuid)",
                        dashboard_id,
                        cid,
                    )

        rows = await pool.fetch(
            """SELECT c.id, c.name, c.slug
               FROM customers c
               JOIN dashboard_assignments da ON da.customer_id = c.id
               WHERE da.dashboard_id = $1::uuid
               ORDER BY c.name ASC""",
            dashboard_id,
        )
        return [dict(r) for r in rows]
    except HTTPException:
        raise
    except asyncpg.InvalidTextRepresentationError:
        raise HTTPException(status_code=400, detail="Invalid ID")
    except Exception as exc:
        raise HTTPException(status_code=500, detail="Internal server error") from exc


@router.get("/{dashboard_id}/customers", dependencies=[Depends(require_auth)])
async def get_dashboard_customers(dashboard_id: str, pool: asyncpg.Pool = Depends(get_pool)):
    # Accept UUID or slug
    is_uuid = bool(UUID_RE.match(dashboard_id))
    try:
        if is_uuid:
            d_id = dashboard_id
        else:
            row = await pool.fetchrow("SELECT id FROM dashboards WHERE slug = $1", dashboard_id)
            if not row:
                raise HTTPException(status_code=404, detail="Dashboard not found")
            d_id = str(row["id"])

        rows = await pool.fetch(
            """SELECT c.id, c.name, c.slug
               FROM customers c
               JOIN dashboard_assignments da ON da.customer_id = c.id
               WHERE da.dashboard_id = $1::uuid
               ORDER BY c.name ASC""",
            d_id,
        )
        return [dict(r) for r in rows]
    except HTTPException:
        raise
    except asyncpg.InvalidTextRepresentationError:
        raise HTTPException(status_code=400, detail="Invalid dashboard ID")
    except Exception as exc:
        raise HTTPException(status_code=500, detail="Internal server error") from exc


@router.delete("/{dashboard_id}", status_code=204, dependencies=[Depends(require_auth)])
async def delete_dashboard(dashboard_id: str, pool: asyncpg.Pool = Depends(get_pool)):
    try:
        result = await pool.execute("DELETE FROM dashboards WHERE id = $1::uuid", dashboard_id)
        if result == "DELETE 0":
            raise HTTPException(status_code=404, detail="Dashboard not found")
    except HTTPException:
        raise
    except asyncpg.ForeignKeyViolationError:
        raise HTTPException(status_code=409, detail="Dashboard is still referenced — unassign any customers first")
    except asyncpg.InvalidTextRepresentationError:
        raise HTTPException(status_code=400, detail="Invalid dashboard ID")
    except Exception as exc:
        raise HTTPException(status_code=500, detail="Internal server error") from exc
